agent-jvm/src/main/java/org/jolokia/mcp/jvmagent/InVmJolokiaService.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jolokia.mcp.jvmagent;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.stream.Collectors;
+import jakarta.enterprise.context.ApplicationScoped;
+
+import org.jolokia.json.JSONObject;
+import org.jolokia.mcp.JolokiaService;
+import org.jolokia.server.core.config.ConfigKey;
+import org.jolokia.server.core.config.Configuration;
+import org.jolokia.server.core.config.StaticConfiguration;
+import org.jolokia.server.core.detector.ServerDetectorLookup;
+import org.jolokia.server.core.request.EmptyResponseException;
+import org.jolokia.server.core.restrictor.RestrictorFactory;
+import org.jolokia.server.core.service.JolokiaServiceManagerFactory;
+import org.jolokia.server.core.service.api.JolokiaContext;
+import org.jolokia.server.core.service.api.JolokiaServiceManager;
+import org.jolokia.server.core.service.api.LogHandler;
+import org.jolokia.server.core.service.api.Restrictor;
+import org.jolokia.server.core.service.impl.CachingServerDetectorLookup;
+import org.jolokia.server.core.service.impl.ClasspathServerDetectorLookup;
+import org.jolokia.server.core.service.impl.ClasspathServiceCreator;
+import org.jolokia.server.core.service.impl.StdoutLogHandler;
+import org.jolokia.server.core.util.ClassUtil;
+import org.jolokia.server.core.util.LocalServiceFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ApplicationScoped
```

Comment:
not being Quarkus expert, just asking here.
So this `InVmJolokiaService` is instantiated by Arc (CDI), but it doesn't do any HTTP request handling.
HTTP is handled by Quarkus MCP through methods (in `org.jolokia.mcp.JolokiaMcpServer`) annotated with `@io.quarkiverse.mcp.server.Tool`?

Reply:
Yes, it's quarkus-mcp-server that hosts a vertx http server to handle and talks MCP protocols via HTTP and uses `JolokiaMcpServer` to register and serve the tools defined there. Then `JolokiaMcpServer` uses either `JolokiaClient` (standalone jolokia-mcp-server) or `inVmJolokiaService` (jolokia-mcp-agent-jvm) to provide the actual Jolokia operations. That's why I thought it's not easy to utilise the existing request handler mechanism for the MCP feature.